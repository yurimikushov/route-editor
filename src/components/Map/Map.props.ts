type MapProps = {
  className?: string
  center?: [lat: number, lon: number]
  zoom?: number
}

export default MapProps
