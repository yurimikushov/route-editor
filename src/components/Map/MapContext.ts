import { createContext } from 'react'
import { Map } from 'yandex-maps'

const MapContext = createContext<Map | null>(null)

export default MapContext
