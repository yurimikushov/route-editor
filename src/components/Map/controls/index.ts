export { default as ZoomControl } from './ZoomControl/ZoomControl'
