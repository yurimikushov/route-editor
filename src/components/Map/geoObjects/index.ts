export { default as Placemark } from './Placemark'
export { default as Polyline } from './Polyline'
