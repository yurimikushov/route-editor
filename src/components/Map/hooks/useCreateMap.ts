import { RefObject, useEffect, useState } from 'react'
import { IMapState, Map } from 'yandex-maps'
import isNull from 'lodash/isNull'
import fetchScript from 'lib/fetchScript'

const MAP_URL = `${process.env.REACT_APP_MAP_API}/?apikey=${process.env.REACT_APP_MAP_API_KEY}`

type Local = 'ru_RU' | 'en_US' | 'en_RU' | 'ru_UA' | 'uk_UA' | 'tr_TR'

const useCreateMap = (
  ref: RefObject<HTMLDivElement>,
  state: IMapState,
  local: Local = 'ru_RU'
) => {
  const [isCreating, setIsCreating] = useState(false)
  const [map, setMap] = useState<Map | null>(null)
  const [error, setError] = useState<Error | null>(null)

  useEffect(() => {
    setIsCreating(true)

    const hash = Date.now().toString(32)
    const loadCallbackName = `load${hash}`
    const errorCallbackName = `error${hash}`

    // @ts-expect-error bad typing
    window[loadCallbackName] = () => {
      if (isNull(ref.current)) {
        throw new Error('[Map] Map container should be created')
      }

      setIsCreating(false)
      setMap(
        new window.ymaps.Map(ref.current, {
          controls: [],
          ...state,
        })
      )

      // @ts-expect-error bad typing
      delete window[loadCallbackName]
      // @ts-expect-error bad typing
      delete window[errorCallbackName]
    }

    // @ts-expect-error bad typing
    window[errorCallbackName] = (error) => {
      setIsCreating(false)
      setError(error)

      // @ts-expect-error bad typing
      delete window[loadCallbackName]
      // @ts-expect-error bad typing
      delete window[errorCallbackName]
    }

    fetchScript(
      `${MAP_URL}&lang=${local}&onload=${loadCallbackName}&onerror=${errorCallbackName}`
    ).catch((error: Error) => {
      setIsCreating(false)
      setError(error)
    })
  }, [])

  return {
    isCreating,
    map,
    error,
  }
}

export default useCreateMap
