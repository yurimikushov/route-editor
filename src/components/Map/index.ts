export { default } from './Map'
export { useMap } from './MapContext'
export { ZoomControl } from './controls'
export { Placemark } from './geoObjects'
