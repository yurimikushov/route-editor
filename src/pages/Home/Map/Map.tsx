import { FC, useMemo, useState } from 'react'
import cn from 'classnames'
import map from 'lodash/map'
import { Point, useRouteEditor } from 'services/routeEditor'
import NativeMap, { Placemark, Polyline, ZoomControl } from 'components/Map'
import MapProps from './Map.props'

const Map: FC<MapProps> = ({ className }) => {
  const { route, changeAddress } = useRouteEditor()
  const [isPointDragging, setIsPointDragging] = useState(false)

  const mapPoints = useMemo(() => {
    return map(route, ({ point }) => point)
  }, [route])

  const handleDragPointStart = () => {
    setIsPointDragging(true)
  }

  const handleDragPointEnd = (id: string, point: Point) => {
    setIsPointDragging(false)
    changeAddress(id, point)
  }

  return (
    <NativeMap className={cn(className, 'relative bg-gray-50')}>
      <ZoomControl className='absolute top-1/2 right-2 md:right-4 transform -translate-y-1/2' />
      {map(route, (address) => (
        <Placemark
          key={address.id}
          point={address.point}
          preset='islands#blueCircleIcon'
          balloonContent={address.name}
          draggable
          onDragStart={handleDragPointStart}
          onDragEnd={(point) => handleDragPointEnd(address.id, point)}
        />
      ))}
      {!isPointDragging && (
        <Polyline
          points={mapPoints}
          strokeWidth={5}
          strokeColor='#1e98fe'
          autoFocusOnLastPoint
        />
      )}
    </NativeMap>
  )
}

export default Map
