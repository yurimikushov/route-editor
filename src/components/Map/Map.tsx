import { FC, useRef } from 'react'
import isNull from 'lodash/isNull'
import Spinner from 'components/Spinner'
import useCreateMap from './hooks/useCreateMap'
import MapContext from './MapContext'
import MapProps from './Map.props'

const MOSCOW = [55.751574, 37.573856]

const Map: FC<MapProps> = ({
  className,
  center = MOSCOW,
  zoom = 9,
  children,
}) => {
  const containerRef = useRef<HTMLDivElement>(null)

  const { isCreating, map, error } = useCreateMap(containerRef, {
    center,
    zoom,
  })

  return (
    <MapContext.Provider value={map}>
      <div className={className}>
        <div
          ref={containerRef}
          className={'w-full h-full flex justify-center items-center'}
        >
          {isCreating && <Spinner className='w-10 h-10' />}
          {!isNull(map) && children}
          {!isNull(error) && (
            <span className='text-xl text-center'>
              Failed to load the map
              <br />
              Try refresh the page
            </span>
          )}
        </div>
      </div>
    </MapContext.Provider>
  )
}

export default Map
