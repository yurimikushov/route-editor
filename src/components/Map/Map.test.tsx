import renderer from 'react-test-renderer'
import { Map as YMap } from 'yandex-maps'
import useCreateMap from './hooks/useCreateMap'
import Map from './Map'

jest.mock('./hooks/useCreateMap')

const runMockUseCreateMap = (result: ReturnType<typeof useCreateMap>) => {
  // @ts-expect-error bad typing
  useCreateMap.mockImplementation(() => result)
}

it('should display spinner during map creation', () => {
  runMockUseCreateMap({
    isCreating: true,
    map: null,
    error: null,
  })

  const tree = renderer.create(<Map />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('should display map when map is created', () => {
  runMockUseCreateMap({
    isCreating: false,
    map: {} as YMap,
    error: null,
  })

  const tree = renderer.create(<Map />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('should display map and passed children when map is created', () => {
  runMockUseCreateMap({
    isCreating: false,
    map: {} as YMap,
    error: null,
  })

  const tree = renderer
    .create(
      <Map>
        <div>Some children</div>
      </Map>
    )
    .toJSON()
  expect(tree).toMatchSnapshot()
})

it('should display error message when map creation is failed', () => {
  runMockUseCreateMap({
    isCreating: false,
    map: null,
    error: new Error(),
  })

  const tree = renderer.create(<Map />).toJSON()
  expect(tree).toMatchSnapshot()
})
