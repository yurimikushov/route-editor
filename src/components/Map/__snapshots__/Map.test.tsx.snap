// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display error message when map creation is failed 1`] = `
<div>
  <div
    className="w-full h-full flex justify-center items-center"
  >
    <span
      className="text-xl text-center"
    >
      Failed to load the map
      <br />
      Try refresh the page
    </span>
  </div>
</div>
`;

exports[`should display map and passed children when map is created 1`] = `
<div>
  <div
    className="w-full h-full flex justify-center items-center"
  >
    <div>
      Some children
    </div>
  </div>
</div>
`;

exports[`should display map when map is created 1`] = `
<div>
  <div
    className="w-full h-full flex justify-center items-center"
  />
</div>
`;

exports[`should display spinner during map creation 1`] = `
<div>
  <div
    className="w-full h-full flex justify-center items-center"
  >
    <svg
      className="w-10 h-10 text-indigo-500 animate-spin"
    >
      spinner.svg
    </svg>
  </div>
</div>
`;
